lib/ecto/exceptions.ex

```diff
@@ -81,22 +81,45 @@ defmodule Ecto.InvalidChangesetError do
   defexception [:action, :changeset]
 
   def message(%{action: action, changeset: changeset}) do
+    changes = extract_changes(changeset)
+    errors = Ecto.Changeset.traverse_errors(changeset, & &1)
+
     """
     could not perform #{action} because changeset is invalid.
 
-    * Changeset changes
+    Changes
+
+    #{pretty changes}
 
-    #{inspect changeset.changes}
+    Params
 
-    * Changeset params
+    #{pretty changeset.params}
 
-    #{inspect changeset.params}
+    Errors
 
-    * Changeset errors
+    #{pretty errors}
 
-    #{inspect changeset.errors}
+    Changeset
+
+    #{pretty changeset}
     """
   end
+
+  defp pretty(term) do
+    inspect(term, pretty: true)
+    |> String.split("\n")
+    |> Enum.map_join("\n", &"    " <> &1)
+  end
+
+  defp extract_changes(%Ecto.Changeset{changes: changes}) do
```

Comment:
The problem is that, when we hide the changeset, we can no longer know if the changeset is going to be inserted/updated/deleted. This way, if something was meant to be deleted but we still include it, it can be confusing. So I thought we should hide the changes not for the main changeset but any of the children changeset that is marked as delete.